File: ui/android/wuffittracker/src/com/androidworks/navsys/wuffit/content/TrackerProvider.java
package com.androidworks.navsys.wuffit.content;

import android.content.ContentProvider;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteQueryBuilder;
import android.net.Uri;
import android.text.TextUtils;

public class TrackerProvider extends ContentProvider {

    private static final String TAG = "TrackerProvider";

    private static final int DETAILS = 1;
    private static final int DETAILS_ITEM = 2;
    private static final int SETUP = 11;
    private static final int SETUP_ITEM = 22;
    private static final int LOCATIONS = 111;
    private static final int LOCATIONS_ITEM = 222;

    private Tracker.DatabaseHelper databaseHelper;

    private static final UriMatcher uriMatcher;

    static {
        uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);
        uriMatcher.addURI(Tracker.AUTHORITY, Tracker.Details.TABLE_NAME, DETAILS);
        uriMatcher.addURI(Tracker.AUTHORITY, Tracker.Details.TABLE_NAME + "/#", DETAILS_ITEM);
        uriMatcher.addURI(Tracker.AUTHORITY, Tracker.Setup.TABLE_NAME, SETUP);
        uriMatcher.addURI(Tracker.AUTHORITY, Tracker.Setup.TABLE_NAME + "/#", SETUP_ITEM);
        uriMatcher.addURI(Tracker.AUTHORITY, Tracker.Locations.TABLE_NAME, LOCATIONS);
        uriMatcher.addURI(Tracker.AUTHORITY, Tracker.Locations.TABLE_NAME + "/#", LOCATIONS_ITEM);
    }


    @Override
    public boolean onCreate() {
        databaseHelper = new Tracker.DatabaseHelper(getContext());
        return true;
    }

    @Override
    public Cursor query(Uri uri, String[] projection, String selection,
                        String[] selectionArgs,
                        String sortOrder) {
        SQLiteQueryBuilder qb = new SQLiteQueryBuilder();

        switch (uriMatcher.match(uri)) {
            case LOCATIONS:
                qb.setTables(Tracker.Locations.TABLE_NAME);
                break;
            case LOCATIONS_ITEM:
                qb.setTables(Tracker.Locations.TABLE_NAME);
                qb.appendWhere(Tracker.Locations._ID + "=" +
                        uri.getPathSegments().get(1));
                break;

            case SETUP:
                qb.setTables(Tracker.Setup.TABLE_NAME);
                break;
            case SETUP_ITEM:
                qb.setTables(Tracker.Setup.TABLE_NAME);
                qb.appendWhere(Tracker.Setup._ID + "=" +
                        uri.getPathSegments().get(1));
                break;

            case DETAILS:
                qb.setTables(Tracker.Details.TABLE_NAME);
                break;
            case DETAILS_ITEM:
                qb.setTables(Tracker.Details.TABLE_NAME);
                qb.appendWhere(Tracker.Details._ID + "=" +
                        uri.getPathSegments().get(1));
                break;

            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }

        // Get the database and run the query
        SQLiteDatabase db = databaseHelper.getReadableDatabase();
        Cursor c = qb.query(db, projection, selection, selectionArgs, null,
                null, sortOrder);

        // Tell the cursor what uri to watch, so it knows when its source data changes
        c.setNotificationUri(getContext().getContentResolver(), uri);
        return c;
    }

    @Override
    public Uri insert(Uri uri, ContentValues initialValues) {
        ContentValues values = new ContentValues(initialValues);
        SQLiteDatabase db = databaseHelper.getWritableDatabase();
        long rowId;
        switch (uriMatcher.match(uri)) {
            case SETUP:
                rowId = db.insert(Tracker.Setup.TABLE_NAME, Tracker.Setup._FID, values);
                if (rowId > 0) {
                    Uri noteUri = ContentUris.withAppendedId(Tracker.Setup.CONTENT_URI, rowId);
                    getContext().getContentResolver().notifyChange(noteUri, null);
                    return noteUri;
                }
                break;
            case LOCATIONS:
                rowId = db.insert(Tracker.Locations.TABLE_NAME, Tracker.Locations._FID, values);
                if (rowId > 0) {
                    Uri noteUri = ContentUris.withAppendedId(Tracker.Locations.CONTENT_URI, rowId);
                    getContext().getContentResolver().notifyChange(noteUri, null);
                    return noteUri;
                }
                break;
            case DETAILS:
                rowId = db.insert(Tracker.Details.TABLE_NAME, Tracker.Details.ID, values);
                if (rowId > 0) {
                    Uri noteUri = ContentUris.withAppendedId(Tracker.Details.CONTENT_URI, rowId);
                    getContext().getContentResolver().notifyChange(noteUri, null);
                    return noteUri;
                }
                break;
            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }
        throw new SQLException("Failed to insert row into " + uri);
    }

    @Override
    public int delete(Uri uri, String where, String[] whereArgs) {
        SQLiteDatabase db = databaseHelper.getWritableDatabase();
        int count;
        switch (uriMatcher.match(uri)) {
            case SETUP:
                count = db.delete(Tracker.Setup.TABLE_NAME, where, whereArgs);
                break;
            case SETUP_ITEM:
                String resultId = uri.getPathSegments().get(1);
                count = db.delete(Tracker.Setup.TABLE_NAME,
                        Tracker.Setup._ID + "=" + resultId
                                + (!TextUtils.isEmpty(where) ?
                                " AND (" + where + ')' :
                                ""), whereArgs);
                break;

            case LOCATIONS:
                count = db.delete(Tracker.Locations.TABLE_NAME, where, whereArgs);
                break;
            case LOCATIONS_ITEM:
                resultId = uri.getPathSegments().get(1);
                count = db.delete(Tracker.Locations.TABLE_NAME,
                        Tracker.Locations._ID + "=" + resultId
                                + (!TextUtils.isEmpty(where) ?
                                " AND (" + where + ')' :
                                ""), whereArgs);
                break;

            case DETAILS:
                count = db.delete(Tracker.Details.TABLE_NAME, where, whereArgs);
                break;
            case DETAILS_ITEM:
                String queryId = uri.getPathSegments().get(1);
                count = db.delete(Tracker.Details.TABLE_NAME,
                        Tracker.Details._ID + "=" + queryId
                                + (!TextUtils.isEmpty(where) ?
                                " AND (" + where + ')' :
                                ""), whereArgs);
                break;

            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }

        getContext().getContentResolver().notifyChange(uri, null);
        return count;
    }

    @Override
    public int update(Uri uri, ContentValues values, String where,
                      String[] whereArgs) {
        SQLiteDatabase db = databaseHelper.getWritableDatabase();
        int count;
        switch (uriMatcher.match(uri)) {
            case DETAILS:
                count = db.update(Tracker.Details.TABLE_NAME, values, where,
                        whereArgs);
                break;
            case DETAILS_ITEM:
                String noteId = uri.getPathSegments().get(1);
                count = db.update(Tracker.Details.TABLE_NAME, values,
                        Tracker.Details._ID + "=" + noteId
                                + (!TextUtils.isEmpty(where) ?
                                " AND (" + where + ')' : ""), whereArgs);
                break;

            case SETUP:
                count = db.update(Tracker.Setup.TABLE_NAME, values, where,
                        whereArgs);
                break;
            case SETUP_ITEM:
                noteId = uri.getPathSegments().get(1);
                count = db.update(Tracker.Setup.TABLE_NAME, values,
                        Tracker.Setup._ID + "=" + noteId
                                + (!TextUtils.isEmpty(where) ?
                                " AND (" + where + ')' : ""), whereArgs);
                break;

            case LOCATIONS:
                count = db.update(Tracker.Locations.TABLE_NAME, values, where,
                        whereArgs);
                break;
            case LOCATIONS_ITEM:
                noteId = uri.getPathSegments().get(1);
                count = db.update(Tracker.Locations.TABLE_NAME, values,
                        Tracker.Locations._ID + "=" + noteId
                                + (!TextUtils.isEmpty(where) ?
                                " AND (" + where + ')' : ""), whereArgs);
                break;

            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }

        getContext().getContentResolver().notifyChange(uri, null);
        return count;
    }

    @Override
    public String getType(Uri uri) {
        switch (uriMatcher.match(uri)) {
            case SETUP:
                return Tracker.Setup.CONTENT_TYPE;
            case SETUP_ITEM:
                return Tracker.Setup.CONTENT_ITEM_TYPE;

            case LOCATIONS:
                return Tracker.Locations.CONTENT_TYPE;
            case LOCATIONS_ITEM:
                return Tracker.Locations.CONTENT_ITEM_TYPE;

            case DETAILS:
                return Tracker.Details.CONTENT_TYPE;
            case DETAILS_ITEM:
                return Tracker.Details.CONTENT_ITEM_TYPE;

            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }
    }

}

File: ui/android/wuffittracker/src/main/kotlin/com/androidworks/navsys/wuffit/content/TrackerProvider.kt
package com.androidworks.navsys.wuffit.content

import android.content.ContentProvider
import android.content.ContentUris
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteQueryBuilder
import android.net.Uri
import android.text.TextUtils

class TrackerProvider : ContentProvider() {

    private var databaseHelper: Tracker.DatabaseHelper? = null

    override fun onCreate(): Boolean {
        databaseHelper = Tracker.DatabaseHelper(context)
        return true
    }

    override fun query(
        uri: Uri,
        projection: Array<String>?,
        selection: String?,
        selectionArgs: Array<String>?,
        sortOrder: String?
    ): Cursor? {
        val qb = SQLiteQueryBuilder()

        when (uriMatcher.match(uri)) {
            LOCATIONS -> qb.tables = Tracker.Locations.TABLE_NAME
            LOCATIONS_ITEM -> {
                qb.tables = Tracker.Locations.TABLE_NAME
                qb.appendWhere("${Tracker.Locations._ID}=${uri.pathSegments[1]}")
            }
            SETUP -> qb.tables = Tracker.Setup.TABLE_NAME
            SETUP_ITEM -> {
                qb.tables = Tracker.Setup.TABLE_NAME
                qb.appendWhere("${Tracker.Setup._ID}=${uri.pathSegments[1]}")
            }
            DETAILS -> qb.tables = Tracker.Details.TABLE_NAME
            DETAILS_ITEM -> {
                qb.tables = Tracker.Details.TABLE_NAME
                qb.appendWhere("${Tracker.Details._ID}=${uri.pathSegments[1]}")
            }
            else -> throw IllegalArgumentException("Unknown URI $uri")
        }

        val db = databaseHelper?.readableDatabase
        val cursor = qb.query(db, projection, selection, selectionArgs, null, null, sortOrder)
        context?.contentResolver?.let { resolver ->
            cursor?.setNotificationUri(resolver, uri)
        }
        return cursor
    }

    override fun insert(uri: Uri, initialValues: ContentValues?): Uri? {
        val values = if (initialValues != null) ContentValues(initialValues) else ContentValues()
        val db = databaseHelper?.writableDatabase
        val rowId: Long

        when (uriMatcher.match(uri)) {
            SETUP -> {
                rowId = db?.insert(Tracker.Setup.TABLE_NAME, Tracker.Setup._FID, values) ?: -1
                if (rowId > 0) {
                    val noteUri = ContentUris.withAppendedId(Tracker.Setup.CONTENT_URI, rowId)
                    context?.contentResolver?.notifyChange(noteUri, null)
                    return noteUri
                }
            }
            LOCATIONS -> {
                rowId = db?.insert(Tracker.Locations.TABLE_NAME, Tracker.Locations._FID, values) ?: -1
                if (rowId > 0) {
                    val noteUri = ContentUris.withAppendedId(Tracker.Locations.CONTENT_URI, rowId)
                    context?.contentResolver?.notifyChange(noteUri, null)
                    return noteUri
                }
            }
            DETAILS -> {
                rowId = db?.insert(Tracker.Details.TABLE_NAME, Tracker.Details.ID, values) ?: -1
                if (rowId > 0) {
                    val noteUri = ContentUris.withAppendedId(Tracker.Details.CONTENT_URI, rowId)
                    context?.contentResolver?.notifyChange(noteUri, null)
                    return noteUri
                }
            }
            else -> throw IllegalArgumentException("Unknown URI $uri")
        }
        throw SQLException("Failed to insert row into $uri")
    }

    override fun delete(uri: Uri, where: String?, whereArgs: Array<String>?): Int {
        val db = databaseHelper?.writableDatabase ?: return 0
        val count: Int

        when (uriMatcher.match(uri)) {
            SETUP -> count = db.delete(Tracker.Setup.TABLE_NAME, where, whereArgs)
            SETUP_ITEM -> {
                val resultId = uri.pathSegments[1]
                val extraWhere = if (!TextUtils.isEmpty(where)) " AND ($where)" else ""
                val selection = "${Tracker.Setup._ID}=$resultId$extraWhere"
                count = db.delete(Tracker.Setup.TABLE_NAME, selection, whereArgs)
            }
            LOCATIONS -> count = db.delete(Tracker.Locations.TABLE_NAME, where, whereArgs)
            LOCATIONS_ITEM -> {
                val resultId = uri.pathSegments[1]
                val extraWhere = if (!TextUtils.isEmpty(where)) " AND ($where)" else ""
                val selection = "${Tracker.Locations._ID}=$resultId$extraWhere"
                count = db.delete(Tracker.Locations.TABLE_NAME, selection, whereArgs)
            }
            DETAILS -> count = db.delete(Tracker.Details.TABLE_NAME, where, whereArgs)
            DETAILS_ITEM -> {
                val queryId = uri.pathSegments[1]
                val extraWhere = if (!TextUtils.isEmpty(where)) " AND ($where)" else ""
                val selection = "${Tracker.Details._ID}=$queryId$extraWhere"
                count = db.delete(Tracker.Details.TABLE_NAME, selection, whereArgs)
            }
            else -> throw IllegalArgumentException("Unknown URI $uri")
        }

        context?.contentResolver?.notifyChange(uri, null)
        return count
    }

    override fun update(
        uri: Uri,
        values: ContentValues?,
        where: String?,
        whereArgs: Array<String>?
    ): Int {
        val db = databaseHelper?.writableDatabase ?: return 0
        val count: Int

        when (uriMatcher.match(uri)) {
            DETAILS -> count = db.update(Tracker.Details.TABLE_NAME, values, where, whereArgs)
            DETAILS_ITEM -> {
                val noteId = uri.pathSegments[1]
                val extraWhere = if (!TextUtils.isEmpty(where)) " AND ($where)" else ""
                val selection = "${Tracker.Details._ID}=$noteId$extraWhere"
                count = db.update(Tracker.Details.TABLE_NAME, values, selection, whereArgs)
            }
            SETUP -> count = db.update(Tracker.Setup.TABLE_NAME, values, where, whereArgs)
            SETUP_ITEM -> {
                val noteId = uri.pathSegments[1]
                val extraWhere = if (!TextUtils.isEmpty(where)) " AND ($where)" else ""
                val selection = "${Tracker.Setup._ID}=$noteId$extraWhere"
                count = db.update(Tracker.Setup.TABLE_NAME, values, selection, whereArgs)
            }
            LOCATIONS -> count = db.update(Tracker.Locations.TABLE_NAME, values, where, whereArgs)
            LOCATIONS_ITEM -> {
                val noteId = uri.pathSegments[1]
                val extraWhere = if (!TextUtils.isEmpty(where)) " AND ($where)" else ""
                val selection = "${Tracker.Locations._ID}=$noteId$extraWhere"
                count = db.update(Tracker.Locations.TABLE_NAME, values, selection, whereArgs)
            }
            else -> throw IllegalArgumentException("Unknown URI $uri")
        }

        context?.contentResolver?.notifyChange(uri, null)
        return count
    }

    override fun getType(uri: Uri): String {
        return when (uriMatcher.match(uri)) {
            SETUP -> Tracker.Setup.CONTENT_TYPE
            SETUP_ITEM -> Tracker.Setup.CONTENT_ITEM_TYPE
            LOCATIONS -> Tracker.Locations.CONTENT_TYPE
            LOCATIONS_ITEM -> Tracker.Locations.CONTENT_ITEM_TYPE
            DETAILS -> Tracker.Details.CONTENT_TYPE
            DETAILS_ITEM -> Tracker.Details.CONTENT_ITEM_TYPE
            else -> throw IllegalArgumentException("Unknown URI $uri")
        }
    }

    companion object {
        private const val DETAILS = 1
        private const val DETAILS_ITEM = 2
        private const val SETUP = 11
        private const val SETUP_ITEM = 22
        private const val LOCATIONS = 111
        private const val LOCATIONS_ITEM = 222

        private val uriMatcher = UriMatcher(UriMatcher.NO_MATCH).apply {
            addURI(Tracker.AUTHORITY, Tracker.Details.TABLE_NAME, DETAILS)
            addURI(Tracker.AUTHORITY, "${Tracker.Details.TABLE_NAME}/#", DETAILS_ITEM)
            addURI(Tracker.AUTHORITY, Tracker.Setup.TABLE_NAME, SETUP)
            addURI(Tracker.AUTHORITY, "${Tracker.Setup.TABLE_NAME}/#", SETUP_ITEM)
            addURI(Tracker.AUTHORITY, Tracker.Locations.TABLE_NAME, LOCATIONS)
            addURI(Tracker.AUTHORITY, "${Tracker.Locations.TABLE_NAME}/#", LOCATIONS_ITEM)
        }
    }
}
